File: 2021/11/two-sum/ruby3/lib/two_sum.rb
#!/usr/bin/env ruby
# typed: true
# frozen_string_literal: true

# https://leetcode.com/problems/two-sum/
#
# Author: Victor Payno (https://github.com/vpayno)

class Solution
  # @return {void}
  def initialize()
    @debug = false
  end

  # @param {::Array[Integer]} nums
  # @param {Integer} target
  # @return {::Array[Integer]}
  def two_sum(nums, target)
    nums_length = nums.length

    for i in 0..(nums_length - 1)
      num1 = nums[i]
      nums_index1 = nums.find_index(num1)

      for j in (nums_index1 + 1)..(nums_length - 1)
        num2 = nums[j]

        test_sum = num1 + num2

        # uncovered
        puts format("%s + %s = %s == %s\n", num1, num2, num1 + num2, target) if @debug

        if test_sum == target
          return [
                   nums_index1,
                   nums[nums_index1 + 1, nums_length].find_index(num2) + nums_index1 + 1,
                 ]
        end
      end
    end

    []
  end # two_sum()

  # @param {::Array[String]} args
  # @return {Integer}
  def main(args)
    search_candidates = if args[0].nil? || args[0].empty?
        []
      else
        args[0].split(" ").reject(&:empty?).map(&:to_i)
      end

    search_target = if args[1].nil? || args[1].empty?
        0
      else
        args[1].to_i
      end

    # uncovered
    $stderr.puts("ERROR: the search candidate list needs at least two numbers in it.\n") if search_candidates.length < 2

    puts format("Search Candidates: %s\n", search_candidates)
    puts format("Searching for sum of %s.\n", search_target)
    puts

    # uncovered
    result = self.two_sum(search_candidates, search_target) if search_candidates && search_target

    if result.length == 2
      puts format("%<num1>s + %<num2>s = %<sum>s\n", num1: search_candidates[result[0]], num2: search_candidates[result[1]], sum: search_target)
      return 0
    else
      print("No solution found.\n")
      return 1
    end
  end # main()
end # class Solution

if $PROGRAM_NAME == __FILE__
  app = Solution.new()  # uncovered
  status = app.main(ARGV)  # uncovered
  exit(status)  # uncovered
end

File: 2021/11/two-sum/python3/two_sum/two_sum.py
#!/usr/bin/env python3
"""
https://leetcode.com/problems/two-sum/

Author: Victor Payno (https://github.com/vpayno)
"""

import sys
from typing import List


class Solution:
    """TwoSum Solution (https://leetcode.com/problems/two-sum/)

    Given an array of integers nums and an integer target, return indices of the two numbers such
    that they add up to target.
    You may assume that each input would have exactly one solution, and you may not use the same
    element twice.
    You can return the answer in any order.
    Only one valid answer exists.
    """

    def __init__(self, debug: bool = False):
        self.debug = debug

    def two_sum(self, nums: List[int], target: int) -> List[int]:
        """Finds two numbers in a list who's sum equals target."""

        compliment: int
        key: int
        value: int

        hashmap = {}

        for key, value in enumerate(nums):
            hashmap[nums[key]] = key

        for key, value in enumerate(nums):
            compliment = target - value

            if compliment in hashmap and hashmap[compliment] != key:
                print("Found solution.")
                return [key, hashmap[compliment]]

        if self.debug:
            print("No solution.")

        return []

    def __str__(self):
        return self.__class__.__name__


if __name__ == "__main__":  # pragma: no cover
    Exercise = Solution()

    search_candidates: List[int] = ([] if sys.argv[1] is None else
                                    [int(_) for _ in sys.argv[1].split(" ")])

    search_target: int = 0 if sys.argv[2] is None else int(sys.argv[2])

    result: List[int] = []

    if len(search_candidates) < 2:
        print(
            "ERROR: the search candidate list needs at least two numbers in it."
        )

    print(f"Search Candidates: {search_candidates}")
    print(f"Searching for sum of {search_target}.")
    print()

    if search_candidates and search_target:
        result = Exercise.two_sum(search_candidates, search_target)

    if len(result) == 2:
        print(f"{result[0]} + {result[1]} = {search_target}\n")
    else:
        print("No solution found.\n")
